Undefined variable when using generic type parameters (PEP695)

### Bug description


The behaviour described in #9110 is still observable with pylint 3.0.2:

```python
def f[T](a: T) -> T: ...
```

works fine but `T` is marked as undefined variable


### Configuration

_No response_

### Command used

```shell
pylint .\mwe.py
```


### Pylint output

```python
************* Module mwe
mwe.py:1:0: C0114: Missing module docstring (missing-module-docstring)
mwe.py:1:0: C0116: Missing function or method docstring (missing-function-docstring)
mwe.py:1:12: E0602: Undefined variable 'T' (undefined-variable)
mwe.py:1:18: E0602: Undefined variable 'T' (undefined-variable)
mwe.py:1:9: W0613: Unused argument 'a' (unused-argument)
mwe.py:1:6: W0612: Unused variable 'T' (unused-variable)
```


### Expected behavior

There should be no E0602, E0602, W0612.

### Pylint version

```shell
pylint 3.0.2
astroid 3.0.1
Python 3.12.0 (tags/v3.12.0:0fb18b0, Oct  2 2023, 13:03:39) [MSC v.1935 64 bit (AMD64)]
```


### OS / Environment

OS: Win32NT (10.0.19045.0)

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/9193.false_positive
+++ b/doc/whatsnew/fragments/9193.false_positive
@@ -1,0 +1,4 @@
+Fix false positives for ``undefined-variable`` and ``unused-argument`` for
+classes and functions using Python 3.12 generic type syntax.
+
+Closes #9193

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1864,12 +1864,6 @@ class VariablesChecker(BaseChecker):
                 or annotation_return
                 or isinstance(defstmt, nodes.Delete)
             ):
-                if (
-                    defined_by_stmt
-                    and isinstance(defstmt, nodes.ClassDef)
-                    and node.parent in defstmt.type_params
-                ):
-                    return (VariableVisitConsumerAction.RETURN, None)
                 if not utils.node_ignores_exception(node, NameError):
                     # Handle postponed evaluation of annotations
                     if not (

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1738,9 +1738,8 @@ class VariablesChecker(BaseChecker):
         elif consumer.scope_type == "function" and self._defined_in_function_definition(
             node, consumer.node
         ):
-            for param in consumer.node.type_params:
-                if node.name == param.name.name:
-                    return False
+            if any(node.name == param.name.name for param in consumer.node.type_params):
+                return False
 
             # If the name node is used as a function default argument's value or as
             # a decorator, then start from the parent frame of the function instead

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1738,6 +1738,10 @@ class VariablesChecker(BaseChecker):
         elif consumer.scope_type == "function" and self._defined_in_function_definition(
             node, consumer.node
         ):
+            for param in consumer.node.type_params:
+                if node.name == param.name.name:
+                    return False
+
             # If the name node is used as a function default argument's value or as
             # a decorator, then start from the parent frame of the function instead
             # of the function frame - and thus open an inner class scope
@@ -1861,6 +1865,12 @@ class VariablesChecker(BaseChecker):
                 or annotation_return
                 or isinstance(defstmt, nodes.Delete)
             ):
+                if (
+                    defined_by_stmt
+                    and isinstance(defstmt, nodes.ClassDef)
+                    and node.parent in defstmt.type_params
+                ):
+                    return (VariableVisitConsumerAction.RETURN, None)
                 if not utils.node_ignores_exception(node, NameError):
                     # Handle postponed evaluation of annotations
                     if not (
@@ -2262,10 +2272,13 @@ class VariablesChecker(BaseChecker):
                     isinstance(defframe, nodes.FunctionDef)
                     and frame is defframe
                     and defframe.parent_of(node)
-                    and stmt is not defstmt
+                    and (
+                        defnode in defframe.type_params
+                        # Single statement function, with the statement on the
+                        # same line as the function definition
+                        or stmt is not defstmt
+                    )
                 ):
-                    # Single statement function, with the statement on the
-                    # same line as the function definition
                     maybe_before_assign = False
                 elif (
                     isinstance(defstmt, NODES_WITH_VALUE_ATTR)

--- a/tests/functional/u/undefined/undefined_variable_py312.py
+++ b/tests/functional/u/undefined/undefined_variable_py312.py
@@ -1,0 +1,7 @@
+# pylint: disable=missing-function-docstring,missing-module-docstring,missing-class-docstring,too-few-public-methods
+
+def f[T](a: T) -> T:
+    print(a)
+
+class ChildClass[T, *Ts, **P]:
+    ...

--- a/tests/functional/u/undefined/undefined_variable_py312.rc
+++ b/tests/functional/u/undefined/undefined_variable_py312.rc
@@ -1,0 +1,2 @@
+[testoptions]
+min_pyver=3.12

